Full documentation of AZSL & AZSLc. (#1158)
o3de/o3de.org: content/docs/atom-guide/dev-guide/shaders/azsl/_index.md

---
linktitle: AZSL

> title: AZSL, The Amazon Shader Language

Author: Fixed.

> description: Learn about the Amazon Shader Language in the Atom Renderer.

Author: Fixed.

toc: true
weight: 100
---

In O3DE, shaders are written in the Amazon Shader Language (AZSL).

> AZSL is a superset of HLSL. AZSL supports Vertex, Fragment and Compute shaders according to Shader Model 6.2. Ray-Tracing shaders are also supported according to Shader Model 6.3.

Author: Fixed.

> AZSL is a superset of HLSL. AZSL supports Vertex, Fragment and Compute shaders according to Shader Model 6.2. Ray-Tracing shaders are also supported according to Shader Model 6.3.

Author: Fixed.

Typically, a shader is written in a file with the *.azsl* extension.

> Shader code can be written in include files with the *.azsli* (optional, yet recommended) extension and referenced by multiple AZSL source files.

Author: Thanks. Fixed.

AZSL is defined on top of HLSL with the following differences:

* [Declaration Of Shader Constants](#shaderconstants).
* [Shader Resource Groups](#ShaderResourceGroup).
* [Shader Resource Group Semantics](srg-semantics).
* [Root Constants](#rootconstant).
* [Shader Variant Options](#option).

> * [The special attribute](#padton) [[pad_to(N)]].

Author: Fixed.

> Although not a language extension, the shader build pipeline allows for *.azsl* files to include other files and define C/C++ Macros. The *.azsl* files are always pre-processed by the *C Pre-Processor (MCPP)* before being transpiled into HLSL. In other words, if a raw *.azsl* file has `#include`, `#pragma` or `#define` directives it should be run through **MCPP** before being compiled with **AZSLc**, otherwise **AZSLc** will fail to compile such files.

Author: Fixed.

> **Caveat**: **MCPP**, doesn't support Macros with arithmetic expressions. Examples:  

Author: I think the styling guide suggests to style in bold the appearance of application names. MCPP is not an acronym .

```cpp
    // Bad (Arithmetic Expression in C-Macros are not supported by MCPP)
    #define BAD_MACRO(x, y) ((x * 0.5) - (y * 2.0))
     
    // Good
    #define HEIGHT (7.2)
```
## <a name="shaderconstants"></a>Declaring Shader Constants (also known as Shader Uniforms in other shader languages, like GLSL)

> AZSL only allows the declaration of shader constants inside the following places:

Author: Fixed.

* [Shader Resource Groups](#ShaderResourceGroup).
* [Root Constants](#rootconstant).

> * `option` variables, but these variables can also be used equivalently as Vulkan's Specialization Constants, in other words, these constants can be statically defined at compilation time to avoid branching at runtime.  

Author: Fixed.

> **AZSLc** will trigger a compilation error if an arbitrary `cbuffer` or any other type of resources are declared outside the scopes mentioned above. The exception would be `static` global variables, which won't be exposed to the application but will be usable internally by the shaders.

Author: Fixed.

## <a name="ShaderResourceGroup"></a>Shader Resource Groups

> A *Shader Resource Group (SRG)* is a logical namespace to declare shader constants and resources. In AZSL the keyword to declare an SRG is `ShaderResourceGroup`. Shader resources are constants of the type Shader Resource Views (SRV), Unordered Access Views (UAV), Constant Buffer Views (CBV), and Sampler States.

Author: Fixed.

> When compiling shaders with the "--use-spaces" argument, each SRG will take over a whole register space (or descriptor set, as known in Vulkan). AZSL abstracts the register and register space assignment of resource descriptors with the `ShaderResourceGroupSemantic` keyword, which must be specified when declaring SRGs.

Author: Fixed.

> A Shader Resource Group is declared with the following syntax:

Author: Fixed.

>     ShaderResourceGroup <Name> : <Semantic>
>     {
>         <Data>
>     };
>      
>     //Alternatively
>     partial ShaderResourceGroup <Name> : <Semantic>
>     {
>         <Data>
>     };

Reviewer: ```suggestion
/```cpp
    ShaderResourceGroup <Name> : <Semantic>
    {
        <Data>
    };
/```

Alternatively, use a partial qualifier to declare an SRG. For more information on partials, refer to [Partial ShaderResourceGroup Definitions](#partial-shaderresourcegroup-definitions).
/```cpp
partial ShaderResourceGroup <Name> : <Semantic>
{
    <Data>
};
/```
```

- Create code blocks by wrapping the code with three ticks \```. In the first set, add `cpp` to enforce code style for C++. You don't need to indent the code block if its wrapped with ```. 
  - NOTE: In the above suggestion, remove the slash in "/```". I only added this to make a suggestion in GitHub comments. 
- Separate the partials example into its own code block and move the description here. Partials introduce a new idea and are worth highlighting. 
- Add a link to the partials section. (This link won't work until you update all anchors.)

Author: Fixed.

  
* **&lt;Name&gt;** is the user defined name for the SRG.
* **&lt;Semantic&gt;** is the name of the `ShaderResourceGroupSemantic` for the SRG. More on this later.

> * **&lt;Data&gt;** is one or more declaration of functions, constants and resources for the binding data contained in the SRG.  

Author: Fixed.

> In the code block shown above, a [partial](#partial) qualifier can be used when declaring SRGs.

Reviewer: ```suggestion
```
- See above suggestion

Author: Fixed.

> ### &lt;Data&gt;

Reviewer: ```suggestion
### Data

SRGs can contain the following data: 
```
- Consider removing the < and > around Data, unless "<Data>" is the actual name of the code element. 
- Note the previous suggestion to reference this section. Update the link if you decide to name this section other than "Data". 
- Add a sentence to describe this section and what customers should do with the information. In this section you list the types of data that SRGs can contain. 
  - Consider using bullets to list these data types, to avoid repeating "SRGs can contain...". (See the following suggestions.)
  Example: 
  
  ![image](https://user-images.githubusercontent.com/75444793/143142754-c395aac0-c648-4a9c-86c7-d6e52378c803.png)

Author: Fixed.

> SRGs can contain fundamental type variables like float, float3, int2, bool, matrices, etc. For each SRG, a Constant Buffer View (CBV) will be automatically created and it will pack all the fundamental type variables declared inside the SRG.

Author: Fixed.

> SRGs can contain resource declarations like Texture2D, Texture2DMS, ByteAddressBuffer, etc. In general any HLSL shader resource of the categories:  Shader Resource Views (SRV), Unordered Access Views (UAV), Samplers and Constant Buffer Views (CBV).

Author: Fixed.

> SRGs can contain unbounded arrays for resource declarations: SRVs, UAVs, CBVs and Samplers. There are some rules that need to be considered, depending on command line arguments when compiling with **AZSLc**. More on unbounded arrays [here](unbounded-arrays).

Author: Fixed.

> SRGs can contain functions.

Reviewer: ```suggestion
- Functions
```

Author: Fixed.

> SRGs can contain structs (declaration & instances). The Shader Build Pipeline and the runtime handle the platform-specific packing rules and alignment.

Reviewer: ```suggestion
- Structs (declaration and instances). The Shader Build Pipeline and runtime handle the struct's platform-specific packing rules and alignment.
```

- Is Shader Buid Pipeline a proper noun? If so, remove "The" before it and capitalize "Shader Build Pipeline" in other places in your docs.
- How does the second sentence relate to structs? Is it the struct's platform-specific packing rules and alignment? I'm not sure what this means and why it's relevant for the user to know this.

Author: Fixed.

> SRGs can contain instances of classes (but you can not declare classes inside an SRG).  

Reviewer: ```suggestion
- Instances of classes. You can instantiate classes in an SRG, but you must declare them outside of the SRG.
```

Author: Fixed.

> **Example**  
> The following example demonstrates declaration of a Shader Resource Group. We have not covered details of what is a `ShaderResourceGroupSemantic`, but suffice to say, it defines the space (aka, descriptor set) index of the constants and resources declared inside the SRG.  

Reviewer: ```suggestion
**Example**  
The following example demonstrates an SRG declaration. A *Shader Resource Group Semantic (SRG Semantic)* defines the space index, or descriptor set, of the constants and resources declared inside the SRG.  When you declare an SRG, it must inherit from an SRG Semantic. For more information on SRG Semantics, refer to [AZSL, Shader Resource Group Semantics](./srg-semantics).
```
- Shorten to "SRG"
- You define SRG Semantic already, so it's not necessary to mention that it hasn't been defined yet. 
- Add a link to the docs for SRG Semantics.
- Consider adding a sentence about the purpose of the SRG Semantic in relation to SRGs. You can change my suggestion to something more technically accurate.
- Verify that "space index" is correct.

Author: Fixed. "it must inherit from an SRG Semantic" changed for "it must specify an SRG Semantic".

>   
>     ShaderResourceGroupSemantic BindingPerExample
>     {
>         FrequencyId = 0;
>     };
>        
>     ShaderResourceGroup ExampleSRG : BindingPerExample
>     {
>         float4          m_uniformColor;
>         Texture2D       m_texture;
>         Buffer<float>   m_buffer;
>         Sampler m_dynamicSampler;
>         Sampler m_staticSampler
>         {
>             AddressU = Wrap;
>             AddressV = Wrap;
>             AddressW = Wrap;
>             MagFilter = Linear;
>         };
>         
>         // You can define functions inside a ShaderResourceGroup
>         float GetRedColor()
>         {
>             return m_uniformColor.r;
>         }
>     };
>   

Reviewer: - Use C++ code style in your code block by wrapping the code with three back ticks (```). (Remove the `\` when you make this change.)
- Remove indent. 
```suggestion
\```cpp  
ShaderResourceGroupSemantic BindingPerExample
{
  FrequencyId = 0;
};
 
ShaderResourceGroup ExampleSRG : BindingPerExample
{
  float4          m_uniformColor;
  Texture2D       m_texture;
  Buffer<float>   m_buffer;
  Sampler m_dynamicSampler;
  Sampler m_staticSampler
  {
      AddressU = Wrap;
      AddressV = Wrap;
      AddressW = Wrap;
      MagFilter = Linear;
  };
  
  // You can define functions inside a ShaderResourceGroup
  float GetRedColor()
  {
      return m_uniformColor.r;
  }
  };
\```  
```

Author: Fixed.

### Scope

> Every SRG defines a scope and the SRG data must be qualified when accessed from out-of-SRG scope code.  
>   
>     float4 color = ExampleSRG::m_uniformColor;
>     color *= ExampleSRG::m_texture.Sample(ExampleSRG::m_staticSampler, float2(0.5, 0.5));
>   

Reviewer: ```suggestion
Every SRG defines a scope. The SRG data must be qualified to access it in code that's out of the SRG's scope.

/```cpp
float4 color = ExampleSRG::m_uniformColor;
color *= ExampleSRG::m_texture.Sample(ExampleSRG::m_staticSampler, float2(0.5, 0.5));
/```
  
```

Remove `/` when making this change.

Author: Fixed.

### Sampler variables

> * Dynamic Samplers are declared without a body definition. They have to be set in the runtime. Dynamic Samplers can also be declared as arrays.
> * Static Samplers are declared with a body definition.
>     * Although you can declare Samplers directly as `SamplerState` or `SamplerComparisonState`, AZSL provides a convenient single `Sampler` keyword for both.
>     * Sampler variables will be compiled into regular HLSL `SamplerState` variables, unless a `ComparisonFunc` is defined inside the `Sampler` block, which will be compiled as `SamplerComparisonState`.  

Reviewer: ```suggestion
A dynamic sampler declaration does not contain a body definition because they must be set at runtime. You can also declare a dynamic sampler as an array.

A static sampler declaration contains a body definition. You can declare a sampler and sampler-comparison by using AZSL's `Sampler` keyword, which conviniently encompasses HLSL's `SamplerState` and `SamplerComparisonState` keywords. AZSLc compiles sampler variables into the corresponding HLSL variables: `SamplerState`, or `SamplerComparisonState`, if it contains a `ComparisonFunc` definition.
```

Author: Fixed.

> <!-- -->

Author: Fixed.

> 
>     //These samplers are declared inside an SRG
>     ...
>     Sampler m_dynamicSampler;
>      
>     // Compiled as HLSL SamplerState
>     Sampler m_staticSamplerState
>     {
>         AddressU = Wrap;
>         AddressV = Wrap;
>         AddressW = Wrap;
>         MagFilter = Linear;
>     };
>      
>     //Compiled as HLSL SamplerComparisonState because of the ComparisonFunc declaration.
>     Sampler m_staticSamplerComparisonState
>     {
>         AddressU = Wrap;
>         AddressV = Wrap;
>         AddressW = Wrap;
>         MagFilter = Linear;
>         ComparisonFunc = Less;
>     };
>     ...

Reviewer: ```suggestion
The following examples demonstrate declarations for a dynamic sampler, a static sampler, and a static sampler comparison. They are declared inside an SRG.

/```c
  // Dynamic sampler declaration.
  Sampler m_dynamicSampler;
   
  // Static sampler declaration. Compiles as HLSL SamplerState.
  Sampler m_staticSamplerState
  {
      AddressU = Wrap;
      AddressV = Wrap;
      AddressW = Wrap;
      MagFilter = Linear;
  };

  // Static sampler declaration. Compiles as HLSL SamplerComparisonState because of the ComparisonFunc definition.
  Sampler m_staticSamplerComparisonState
  {
      AddressU = Wrap;
      AddressV = Wrap;
      AddressW = Wrap;
      MagFilter = Linear;
      ComparisonFunc = Less;
  };
/```
```

Author: Fixed.

  
### Functions

> Functions can be declared in classes, global scopes, or inside SRGs. Declaring functions in classes or in the global scope is a standard HLSL feature. Defining functions inside SRGs is unique to AZSL (Of course, because ShaderResourceGroup  is a exclusive feature of AZSL).  

Author: Fixed.

>   

Reviewer: ```suggestion

The following example demonstrates function declarations inside and outside an SRG. 

```

>     ShaderResourceGroup PerObject : BindingPerObject

Reviewer: For the whole code block
- Unindent
- Wrap in three back ticks (```) with `c` syntax highlighting.

Author: Fixed.

        {
            float4x4 modelToWorld = float4x4(
                float4(1, 0, 0, 0),
                float4(0, 1, 0, 0),
                float4(0, 0, 1, 0),
                float4(0, 0, 0, 1));
       
            modelToWorld[0] = m_modelToWorld[0];
            modelToWorld[1] = m_modelToWorld[1];
            modelToWorld[2] = m_modelToWorld[2];
            return modelToWorld;
        }
           
        float3x3 GetNormalMatrix()
        {
            return m_normalMatrix;
        }
    }
      
      
    VertexOutput MainVS(VertexInput input)
    {
        const float4x4 worldMatrix = PerObject::GetWorldMatrix();
       
        VertexOutput output;
        float3 worldPosition = mul(worldMatrix, float4(input.m_position,1)).xyz;
        output.m_position = mul(GetView_ViewProjectionMatrix(), float4(worldPosition, 1.0));
       
        output.m_uv = input.m_uv;
               
        output.m_positionToCamera = GetView_WorldPosition() - worldPosition;
       
        output.m_normal = mul(PerObject::GetNormalMatrix(), input.m_normal);
        output.m_tangent = mul(PerObject::GetNormalMatrix(), input.m_tangent);
       
        return output;
    }
  
### Structs

> Just like in HLSL, structs can be defined globally, but in AZSL they can also be defined and instantiated inside SRGs. You can only instantiate globally defined structs inside SRGs, rootconstants or inside another struct or class.  
>   
> In general, structs are supported just like in HLSL. struct has no constructors/destructors and as such, do not accept default member initializers.  
>   
> Just like in HLSL, struct has the particularity to be data only (no methods).

Reviewer: - I think we can tighten the comparison between HLSL and AZSL by making separate lists of the rules that HLSL brings and of the additional rule(s) that AZSL brings. 

- Add space between "root constants", as that's how they are written in [Direct3D docs]( https://docs.microsoft.com/en-us/windows/win32/direct3d12/using-constants-directly-in-the-root-signature). 
```suggestion
AZSL supports structs similarly as in HLSL and adds extended rules. 

Just like in HLSL, the following rules apply for structs:
- Can be defined globally. 
- Contains only data, not methods. 
- Don't have a constructor or destructor. 
- Don't accept default member initializers.

AZSL adds an extended rule: Define and instantiate structs inside SRGS. You can instantiate globally defined structs inside SRGs, root constants, and another struct or class. 
```

Author: Fixed.

### Classes
Classes are supported in AZSL with the exact same expectations of how they are supported in HLSL. They can only be declared/defined in global scopes (outside of SRGs), but an SRG can instantiate class type variables.  
  
Like in HLSL, classes can inherit from interfaces to be forced to respect a minimal method set.  
  
Unlike in HLSL, AZSL forbids class inheritance among classes.  
  

> Classes has no constructors/destructors and as such, do not accept default member initializers.

Author: Fixed

### Matrices
AZSL follows the HLSL convention to use column major matrices by default.  
  
However, in the O3DE Runtime, the default convention is row major matrices.  
  

> When declaring a matrix, the order can be specified using the row_major or column_major keywords.  

Author: fixed.

  
    ShaderResourceGroupSemantic MatrixExample
    {
        FrequencyId = 1;
    };
      
    ShaderResourceGroup MatrixExampleSRG : MatrixExample
    {
                     float2x3 defaultMatrix;
        row_major    float2x3 rowMajorMatrix;
        column_major float2x3 colMajorMatrix;
    };
  
### Arrays And Unbounded Arrays

> From the point of view of defining arrays, the rules and limitations are exactly the same as HLSL. Examples:  

Author: Fixed.

  
    ShaderResourceGroup PerObject : BindingPerObject
    {
        float m_arrayOfFloats[64]; // GOOD. Arrays of fundamental types must be of constant size.
     
        struct SomeStruct
        {
            int m_integer;
            bool m_boolean;
        }
        static const int MyArraySize = 32;
        SomeStruct m_arrayOfStructs[MyArraySize]; // GOOD. Arrays of packable types (structs & fundamentals) require constant size definition.
     
        int m_unboundedArrayOfIntegers[]; // ERROR. Unbounded arrays are not supported for fundamental(packables, in general) data types.
     
        Texture2D<float4> m_textures[MyArraySize]; // GOOD. Contant size array of Resource View types is ok.
         
        ConstantBuffer<SomeStruct> m_unboundedArrayOfStructs[]; // GOOD. Unbounded array of Resource View type is ok. BUT, there are some restrictions.
    }
  

> To learn more about what are the limitations when declaring unbounded arrays, go here: [Binding Rules For Unbounded Arrays.](unbounded-arrays)  

Reviewer: ```suggestion
To learn more about the limitations when declaring unbounded arrays, go to [Binding Rules For Unbounded Arrays](unbounded-arrays).
```

Author: Fixed.

> ### <a name="partial"></a> Partial ShaderResourceGroup Definitions.

Reviewer: - If you make this fix, please fix references to this section. For example:
   - Within the same page: [Partial Shader Resource Groups](#partial-shader-resource-groups)
   - Outside of this page: [Partial Shader Resource Groups](<path>/azsl/_index.md#partial-shader-resource-groups)
```suggestion
### Partial Shader Resource Groups
```

Author: Fixed.

> Before talking about `partial ShaderResourceGroup`, it's important to talk about [ShaderResourceGroupSemantic](srg-semantics), which is always required when declaring a `ShaderResourceGroup`.  
> Go to this section to read about [ShaderResourceGroupSemantic](srg-semantics).  

Reviewer: - Only need one link to the docs. 
- Switch to "partial SRGs" when introducing the concept. Later, use `partial ShaderResourceGroup` in code style when talking about defining them in code. 
```suggestion
Before talking about partial SRGs, it's important to talk about SRG Semantics, which is always required when declaring a `ShaderResourceGroup`. Read about SRG Semantics in [Shader Resource Group Semantics](srg-semantics).
```

Author: Fixed.

> The `partial` keyword is useful to define parts of an SRG in different blocks. For example, across different files. This is a similar concept as partial classes in C#.  
> There are no limits into how many `partial` blocks can be used to define an SRG.  
> There are only three requirements:  
> 1. When using `partial` to define an SRG block, all other SRG definition blocks, for the same SRG, must start with the `partial` keyword.
> 2. For a given SRG, at least one `partial` SRG block must define the Shader Resource Group Semantic.
> 3. For a given SRG, if the Shader Resource Group Semantic is defined more than once across different `partial` blocks, they must refer to the same `ShaderResourceGroupSemantic`.  

Reviewer: - Added the first sentence to clarify the code used to define partial SRGs.
```suggestion
Define a partial SRG by using `partial ShaderResourceGroup`. The `partial` keyword is useful to define parts of an SRG in different blocks (for example, across different files). This is a similar concept as partial classes in C#.  There are no limits on how many `partial` blocks define an SRG.

There are three requirements for defining partial SRGs:
1. When using `partial` to define an SRG block, all other SRG definition blocks within the same SRG must also start with the `partial` keyword.  
2. For a given SRG, at least one partial SRG block must define the SRG Semantic.  
3. For a given SRG, if the SRG Semantic is defined more than once across different partial blocks, then they must refer to the same `ShaderResourceGroupSemantic` object.   
```

Author: Fixed.

>   

Reviewer: - I suggest moving the remark you wrote below here, before the examples.
  Original text:
  > An important remark about partial SRG blocks is that when the compiler, **AZSLc**, finds the first `partial` block, it will use it as the unified point of emission for all the data of a given SRG across all `partial` blocks. The main recommendation that comes out of this rule, is that any globally defined `struct`, or `class` that may be referenced by a `partial` SRG block must be defined before the first partial block for the SRG in question.  

- If it's important, you can use {{< note >}} ... {{< /note>}} to create a Note callout box. To see this, preview the docs in a website preview. 

- Specify "SRG" when talking about "partial SRG".
- Don't use in-line code style for words such as classes, structs, partials, unless you are specifically referring to the keyword. For example: "Use the `partial` keyword to create partial blocks."

```suggestion
{{< note >}}
During compilation, **AZSLc** uses the first partial SRG block it finds as the unified point of emission for all the data of a given SRG across all partial SRG blocks. Therefore, we recommend that you define structs and classes that are referenced by a partial SRG before the first partial block.
{{< /note >}}
```

Author: Fixed

**Example: Using the partial keyword**  
  
    ShaderResourceGroupSemantic PerPass
    {
        FrequencyId = 1;
    }
     
    partial ShaderResourceGroup PassSrg
    {
        Texture2D m_texture; // A regular SRV resource,
    }
     
    partial ShaderResourceGroup PassSrg : PerPass // At least one partial block must define the semantic
    {
        Sampler m_sampler; // A Dynamic sampler.
    }
     
    partial ShaderResourceGroup PassSrg
    {
        float2 m_resolution;
        float4 ReadColor(float2 uv)
        {
            return m_texture.Sample(m_sampler, uv);
        }
    }
     
    PSOutput MainPS(VSOutput IN)
    {
        PSOutput OUT;
        float2 uv = IN.m_position.xy / PassSrg::m_resolution;
        OUT.m_color = PassSrg::ReadColor(uv);
        return OUT;
    }
  

> An important remark about partial SRG blocks is that when the compiler, **AZSLc**, finds the first `partial` block, it will use it as the unified point of emission for all the data of a given SRG across all `partial` blocks. The main recommendation that comes out of this rule, is that any globally defined `struct`, or `class` that may be referenced by a `partial` SRG block must be defined before the first partial block for the SRG in question.  

Reviewer: See suggestion above. Move this before the first example, so it doesn't get lost amidst the series of examples.
```suggestion
```

Author: Fixed.

---
linktitle: AZSL
title: AZSL, The Amazon Shading Language
description: Learn about the Amazon Shading Language (AZSL) for Atom Renderer.
toc: true
weight: 100
---

> Shaders for Atom Renderer are written in Amazon Shading Language (AZSL). AZSL is a superset of high-level shader language (HLSL). AZSL supports Vertex, Fragment and Compute shaders according to Shader Model 6.2. Ray-Tracing shaders are also supported according to Shader Model 6.3.

Author: Fixed.

Typically, you write shader code in a file with the `.azsl` extension, also known as an _AZSL source file_. You can also write shader code in a file with the `.azsli` extension, which is an include file that multiple AZSL source files can reference. Consider writing your shaders in `.azsli` files to reuse them in multiple AZSL source files and reduce code redundancy.  

AZSL is defined on top of HLSL with the following differences:

* [Declaration Of Shader Constants](#declaring-shader-constants)
* [Shader Resource Groups](#ShaderResourceGroup)
* [Shader Resource Group Semantics](srg-semantics)
* [Root Constants](#root-constants)
* [Shader variant options](#option)
* [The special attribute \[\[pad_to(N)\]\]](#padton)

> The shader build pipeline allows for `.azsl` files to include other files and define C/C++ Macros. The `.azsl` files are always pre-processed by the **[C Pre-Processor (MCPP)](http://mcpp.sourceforge.net/)** before being transpiled into HLSL. In other words, if a raw `.azsl` file has `#include`, `#pragma` or `#define` directives it should be run through **MCPP** before being compiled with **AZSLc**, otherwise **AZSLc** will fail to compile such files.

Author: Fixed.

> **Caveat**: **MCPP** doesn't support Macros with arithmetic expressions. Examples:  

Author: What about the styling guide to always bold application  names?

> **Caveat**: **MCPP** doesn't support Macros with arithmetic expressions. Examples:  

Reviewer: You only need to bold the first instance of the application name on the page. So you bolded **[C Pre-Processor (MCPP)](http://mcpp.sourceforge.net/)** in the previous paragraph. In subsequent instances of MCPP, such as here, you don't need to bold it. 
Check out https://o3de.org/docs/contributing/to-docs/style-guide/format/#bold-proper-application-names.

> **Caveat**: **MCPP** doesn't support Macros with arithmetic expressions. Examples:  

Author: Fixed.

```cpp
    // Bad (Arithmetic Expression in C-Macros are not supported by MCPP)
    #define BAD_MACRO(x, y) ((x * 0.5) - (y * 2.0))
     
    // Good
    #define HEIGHT (7.2)
```

> ## Declaring Shader Constants

Author: Fixed.

> Also known as Shader Uniforms in other shader languages, like GLSL.

Author: Fixed.

AZSL allows the declaration of shader constants only in the following places:

> * [Shader Resource Groups](#ShaderResourceGroup).

Author: Fixed.

> * [Root Constants](#root-constant).

Author: Fixed

* `option` variables
{{< note >}}
You can also use `option` variables equivalently as Vulkan's Specialization Constants. In other words, You can define these constants statically at compilation time to avoid branching at runtime.

> {{< /note >}}

Author: Done.

> **AZSLc** triggers a compilation error if an arbitrary `cbuffer` or any other type of resource is declared outside of the scopes mentioned above. The exception is `static` global variables, which shaders use internally, but aren't exposed to the application.

Author: Fixed.

> ## <a name="ShaderResourceGroup"></a>Shader Resource Groups

Author: Done.

> A *Shader Resource Group (SRG)* is a logical namespace to declare shader constants and resources. In AZSL the keyword to declare an SRG is `ShaderResourceGroup`. Shader resources are constants of the types: Shader Resource Views (SRV), Unordered Access Views (UAV), Constant Buffer Views (CBV), and Sampler States.

Reviewer: ```suggestion
A *Shader Resource Group (SRG)* is a logical namespace to declare shader constants and resources. In AZSL, the keyword to declare an SRG is `ShaderResourceGroup`. Shader resources are constants of the types: Shader Resource Views (SRV), Unordered Access Views (UAV), Constant Buffer Views (CBV), and Sampler States.
```

Author: Fixed.

            return m_uniformColor.r;
        }
    };
```
### Scope

> Every SRG defines a scope and the SRG data must be qualified when accessed from out-of-SRG scope code.  

Reviewer: - Improve flow.
- Be consistent with "**Example**"

```suggestion
Every SRG defines a scope. The SRG data must be qualified when accessed from outside of the SRG's scope in code.  

**Example**:
```

Author: Fixed.

```cpp
    float4 color = ExampleSRG::m_uniformColor;
    color *= ExampleSRG::m_texture.Sample(ExampleSRG::m_staticSampler, float2(0.5, 0.5));
```

### Sampler variables
A dynamic sampler declaration does not contain a body definition because they must be set at runtime. You can also declare a dynamic sampler as an array.

A static sampler declaration contains a body definition. You can declare a sampler and sampler-comparison by using AZSL's `Sampler` keyword, which conveniently encompasses HLSL's `SamplerState` and `SamplerComparisonState` keywords. AZSLc compiles sampler variables into the corresponding HLSL variables: `SamplerState`, or `SamplerComparisonState`, if it contains a `ComparisonFunc` definition..  

> The following examples demonstrate declarations for a dynamic sampler, a static sampler, and a static sampler comparison. They must be declared inside an SRG.

Reviewer: For consistency with "**Example**".
```suggestion
**Example**: 

The following examples demonstrate declarations for a dynamic sampler, a static sampler, and a static sampler comparison. They must be declared inside an SRG.
```

Author: Fixed.

        ComparisonFunc = Less;
    };
    ...
```
### Functions

> Functions can be declared in classes, global scopes, or inside SRGs. Defining functions inside SRGs is unique to AZSL.  

Reviewer: ```suggestion
Functions can be declared in classes, global scopes, or inside SRGs. Defining functions inside SRGs is unique to AZSL.  

**Example**:

```

Author: Fixed.

       
        return output;
    }
```
### Structs
Just like in HLSL, the following rules apply for structs:
- Must be defined globally. 
- Contains only data, not methods. 
- Cannot have a constructor or destructor. 
- Cannot accept default member initializers.  
  
AZSL adds an extended rule: Define and instantiate structs inside SRGS. You can instantiate globally defined structs inside SRGs, root constants, and another struct or class.

### Classes
Just like in HLSL, the following rules apply for classes:
- Must be declared and defined only in global scopes, outside of SRGs. 
- Can inherit from interfaces, which forces the class to repect a minimal method set. 
- Don't have constructors or destructors.
- Don't accept default member initializers.  
  
ASZL adds the following extended rules for classes:
- Can instantiate class type variables in an SRG. 
- Cannot inherit classes, unlike in HLSL. .

### Matrices

> AZSL follows the HLSL convention to use column major matrices by default. However, in O3DE runtime, the default convention is row major matrices. When declaring a matrix, you can specify the order by using the `row_major` or `column_major` keywords.  

Reviewer: ```suggestion
AZSL follows the HLSL convention to use column major matrices by default. However, in O3DE runtime, the default convention is row major matrices. When declaring a matrix, you can specify the order by using the `row_major` or `column_major` keywords.  

**Example**:

```

Author: Done.

        OUT.m_color = PassSrg::GetColor();
        return OUT;
    }
```
## Root Constants

> AZSL supports the definition of shader constants that conform to the DX12 concept of Root Constants: https://docs.microsoft.com/en-us/windows/win32/direct3d12/using-constants-directly-in-the-root-signature.  

Reviewer: ```suggestion
AZSL supports the definition of shader constants that conform to [DirectX 12's concept for root constants](https://docs.microsoft.com/en-us/windows/win32/direct3d12/using-constants-directly-in-the-root-signature).
```

Author: Fixed.

> Root Constants, declared with the `rootconstant` keyword, will be transparently embedded into a `ConstantBuffer`. They are useful to define shader constants of frequent access. Because they are located in the root signature they can be read instantly without requiring additional levels of indirection.  

Reviewer: Can you verify the suggestion is accurate? Is AZSLc responsible for processing root constants as described here?
```suggestion
Root constants are declared with the `rootconstant` keyword. AZSLc transparently embeds root constants into a `ConstantBuffer`. They are useful to define shader constants that you frequently access. Because they are located in the root signature, AZSLc reads them instantly without requiring additional levels of indirection.  
```

Author: The only thing wrong is saying "AZSLc reads them". When the shader code runs inside the GPU it is the shader code itself who reads those constants.

Reviewer: Okay sounds good. You can ignore the "AZSLc reads them" suggestion.

> Shader constants qualified as `rootconstant` are always declared globally. Examples:  

Reviewer: For consistency with "**Example**"
```suggestion
Shader constants qualified as `rootconstant` are always declared globally. 

**Example**:  
```

Author: Fixed.

```cpp
    rootconstant float4x4 s_objectMatrix;
    rootconstant uint s_materialIndex;
```

> Root Constants should be used sparingly because the space to define root constants, the Root Signature, is shared with Root Descriptors and Descriptor Tables. In D3D12, the Root Signature is limited at 64 DWords (256 Bytes) worth of space that is shared among Root Constants, Root Descriptors and Descriptor Tables.  

Reviewer: - Switch to active voice.
- Improve flow. 
```suggestion
Consider using root constants sparingly due to limitations with root signatures, the space to define root constants. Root signatures are shared between root constants, root descriptors, and descriptor tables. In Direct3D 12, a root signature is limited to 64 DWords (256 bytes of space). 
```

Author: Fixed.

> ## <a name="option"></a>Shader variant options

Reviewer: Fix header as described previous comment.

Author: Fixed.

> Before describing how to use Shader variant options in a shader, you should read this: [AZSL: Design Principles And Purpose Of Shader variant options.](shader-variant-options)  

Reviewer: Is it necessary for users to read **AZSL: Design Principles And Purpose Of Shader variant options** before reading this section? 

This is the "overview" page of the AZSL docs, so I believe that users should get a glimpse of what shader variant options are without needing to read the other docs. Instead of requiring them to read the other docs first, I suggest briefly defining shader variant options in this section. Then, refer them to the other docs for more information. 
```suggestion

```

Author: Fixed.

  
Shader variant options are declared as global variables with the `option` keyword. Their ultimate purpose is to avoid conditional branching, in favor of static execution, all for the sake of performance.  
  

> Although the ultimate purpose is to avoid conditional branching at runtime, they can be used to conditionally branch at runtime and, in fact, it's the default behavior of the shader code to conditionally branch at runtime when a Shader Variant has not been compiled (not *baked*).  

Reviewer: 1. I'm unsure about how to use Shader variant options based on this description. It says that shader variant options are used to avoid conditional branching, but that its default behavior is to conditionally branch. Does that mean we need to "enable" stader variant options somehow? 

    My assumption is that shader constants conditionally branch, and to avoid that we define them with `option` keyword. This defines them as shader variant options, which avoid branching at conditional statements. **Can you verify that the suggestion is correct and make adjustments if necessary?**

2. Add "For more info..." if you decide to remove the first sentence, "Before describing....". 

```suggestion
Shader variant options are shader constants that are optimized for conditional statements. By default, shader constants conditionally branch at runtime. To avoid conditional branching and improve performance, Shader variant options use static execution instead of conditional branching. For more information on shader variant options, refer to [AZSL: Design Principles And Purpose Of Shader variant options.](shader-variant-options).
```

Author: Changed it all to this:

Shader variant options are declared as global variables with the `option` keyword. Their ultimate purpose is to be used as statically defined flags. It is helpful to think of Shader variant options as compile time configurable C Macros.  
By default, during shader asset compilation, all shader variant options are compiled as global shader constant flags that can be used for conditional branching. For the sake of performance, You can define the values of each shader variant option at compilation time, and thanks to shader compiler optimizations the conditional branches are removed in favor of static execution of some sections of the code.

> It is helpful to think of Shader variant options as compile time configurable C Macros.  

Reviewer: This sentence alone doesn't give much information. Either remove it, or explain in what way it is related to compile time configurable C macros. 
```suggestion
```

Author: Moved it.

> They can be of type `bool`, `int`, or `enum`. Other data types like `float` and `struct` options are not supported.  

Reviewer: Consider teaching users how to declare shader variant options here, where it's closer to the example and other data types. I believe this is a better place to learn about `option` keyword, because the previous paragraphs are more about what shader variant options are, and this is about how to use them. 

```suggestion
Declare Shader variant options as global variables with the `option` keyword. Shader variant options can be of type `bool`, `int`, or `enum`. They cannot be of type `float` or `struct`.
```

Author: Fixed.

> An `int` option requires a range attribute to specify the minimum and maximum range of values.  

Reviewer: - Tighten sentence and switch to active voice. 
```suggestion
 For `int` types, specify the inclusive minimum and maximum range of values. 
```

Author: Fixed.

     
    [[range(3, 16)]] // This integer option accepts values between 3 and 16 (both ends included).  
    option int o_numberOfTaps;
```

> Even though the Shader variant options are declared globally, they are actually encoded in a single array of bits as a member variable of one, and only one, SRG.  

Reviewer: Switch to singular "a Shader variant option" to clarify that each shader variant option is encoded in a single array, rather than all collectively. 
Other suggestions to improve flow and clarify details.
```suggestion
Even though a Shader variant options is global, it's encoded in a single array of bits because it's a member of only the SRG that defines it. To learn more about how AZSLc encodes Shader variant options during compilation, read [Shader variant options & The Fallback Key](shader-variants-fallback-key).
```

Author: The rewording is not accurate. Leaving as is.

To learn more about how Shader variant options are encoded when compiled, please read: [Shader variant options & The Fallback Key.](shader-variants-fallback-key)  
  

> # <a name="padton"></a>The special attribute [[pad_to(N)]]

Reviewer: Fix header as described previously.

Author: What would be the encoding for this anchor?
"# The special attribute \[\[pad_to(N)\]\]

Reviewer: #the-special-attribute-pad_ton

In case you're wondering how to find this out, I have a few tools that make it easy. 
1. The **Markdown All in One** extension for VS Code. It has an autocomplete feature and told me the link: 
![image](https://user-images.githubusercontent.com/75444793/144963021-bfc23540-750b-4979-b69b-b0c6a229d881.png)
2. When I preview the website on my local machine (requires that you install Hugo), I navigate to the anchor icon by the section name, and click on the URL. The encoding is written in the browser URL: 
![image](https://user-images.githubusercontent.com/75444793/144963285-66bb4b33-df41-47ab-a10d-75771a0e388d.png)

> In DX12, the layout, offset & sizes, of variables inside `struct` definitions changes whether a `struct` is being used as part of a `ConstantBuffer`, or as part of a `StructuredBuffer`.  

Reviewer: Add context to clarify whether or not this is a problem that the special attribute tries to resolve and when the problem occurs (e.g. during compilation). 
```suggestion
When AZSLc compiles a variable defined inside structs, the `[[pad_to(N)]]` attribute helps guarantee the variable's alignment, which is a problem that occurs in DX12. In DX12, for variables inside `struct` definitions, the variable's layout, offset, and size changes depending on if the struct is part of a `ConstantBuffer` or a `StructuredBuffer`.  
```

Author: Fixed.

> Example:  

Reviewer: ```suggestion
**Example**:

```

Author: Fixed.

>   

Reviewer: > For ConstantBuffer&lt;MyStruct&gt; case you'll get these offsets:  
> ...
> For StructuredBuffer&lt;MyStruct&gt; case you'll get:  
> ...

These don't need to be in a code block. Consider the suggestions below to remove them from the code block and list the offsets in bullet points. 

1) Make the following edits for lines 438 to 442:
   - Close the code block (remove `/` when you copy/paste).
```
/```

`ConstantBuffer<MyStruct>` produces the following offsets:  

- `float m_data`: Offset = 0
- `float4 m_arr[2]`: Offset = 16
```

2) Make the following edits for lines 444 to 447:

```
`StructuredBuffer<MyStruct>` produces the following offsets:  

- `float m_data`: Offset = 0
- `float4 m_arr[2]`: Offset = 4
```

Author: Fixed.

    float m_data;                     ; Offset:    0
    float4 m_arr[2];                  ; Offset:    4
```

> The [[pad_to(N)]] attribute is useful to pad the data with dummy variables until guaranteeing that the offset of the next variable starts with the desired alignment. The **N** argument is a single integral literal that must be a multiple of 4.  

Reviewer: ```suggestion
The `[[pad_to(N)]]` attribute pads the data with dummy variables to guarantee that the offset of the next variable starts with the desired alignment. The **N** argument is a single integral literal that must be a multiple of 4. 
```

Author: Fixed.

> In the case posted above, `struct MyStruct`, using `[[pad_to(16)]]` can be useful to guarantee the same layout regardless of how it is used, as part of a `ConstantBuffer` or a `StructuredBuffer`:  

Reviewer: ```suggestion

To resolve the example above, use `[[pad_to(N)]]` attributes to pad the data. This guarantees that `struct MyStruct` has the same layout, regardless of whether its a `ConstantBuffer` or `StructuredBuffer`.
```

Author: Fixed.

> ```cpp

Reviewer: You can combine the last two code blocks into 1. 

Make the following edits to lines 454 to 456:
```
         float m_data;
         [[pad_to(16)]]       // This guarantees that the variable starts at offset 16.
         float4 m_arr[2];   
         [[pad_to(64)]]      // This guarantees that sizeof(MyStruct) is 64.
```

Author: I think it's important to keep them separate, even though they look similar.

    struct MyStruct
    {
         float m_data;
         [[pad_to(16)]]
         float4 m_arr[2]; //Now, this member variable will always start at offset 16.
    };
```
In this example, **MyStruct** is forced to be of size 64 bytes:  
```cpp
    struct MyStruct
    {
         float m_data;
         [[pad_to(16)]]
         float4 m_arr[2]; //Now, this member variable will always start at offset 16.
         [[pad_to(64)]] // This guarantees the sizeof(MyStruct) to be 64.
    };

> ```

Reviewer: Remove this if you combined the examples into one, as shown in the previous suggestion.

Author: Kept separated.
